packages/@glimmer/runtime/lib/vm/element-builder.ts

```diff
@@ -237,6 +262,15 @@ export class NewElementBuilder implements ElementBuilder {
   }
 
   popRemoteElement() {
+    if (this.env.debugRenderTree) {
+      const state = this.remoteStateStack.pop();
+      const element = this.element;
+      this.env.debugRenderTree?.didRender(state!, {
+        parentElement: () => (element as any).parentElement,
+        firstNode: () => element,
+        lastNode: () => element,
+      });
+    }
```

Comment:
`this.popBlock()` returns the bounds so you would just call it first, store it in a local variable, and then pass that to `didRender`, rather than having to make your own

Reply:
As for "state", it just have to be a stable object that uniquely identify this node, and so it actually works to just use the "block" object – the one you get back from calling `pushLiveBlock` and `popBlock` (regarding my comment above, `Block implements Bounds`), which would be the same object, because that is already stored in a stack, which is why I said it would be unnecessary for you to maintain yet another parallel stack just to track the same.

However, it looks like something is currently wrong with the types so you would have to fix it before you can do that.

First of all, you should have access to the block returned by `pushLiveBlock` (so you can pass it to `debugRenderTree.create` from within `pushRemoteElement`). However, that is happening from with `__pushRemoteElement`, which currently has a return type of `Nullable<RemoteLiveBlock>`. I think that is wrong – specifically, the `Nullable<...>` is not accurate, as far as I can tell, there is no code path that would return `null`, and if there is, it would be a bug to unconditionally call `popBlock` here in `popRemoteElement`.

So you'll have to fix that return type, and then do the same in the rehydration builder, then, you should have everything you need to implement the feature without having to add your own parallel stack.

Reply:
And – while you are in this file, we really shouldn't be inferring the return types of these methods in this file (ie. a bunch of methods does not have `: ...` here). It seems like most of them are just `: void` and we should have typed that explicitly.

Ideally we should fix that in a separate PR, either just in this file, or across the whole codebase and enforce it with a lint, that would be great. It will probably give @NullVoxPopuli some rebase nightmares though, so you may want to coordinate with them on Discord about how to approach it; I think it is ultimately a rather mechanical fix that should be possible to do, but @NullVoxPopuli can figure that out with you.

Reply:
block seems to have the bounds set to the children of the remote element. so I reverted to pass custom bounds

Reply:
That is the correct usage of bounds though. It is what each node has rendered. A component’s bounds is its content. And you need that if you want to draw the bounding box if the inspector

Reply:
True, it will not work with insertAfter, otherwise this would work. 
Would it be okay to still have custom bounds where parentElement is the parent of the in-element? I would like to have that link for the inspector.
 first&last would use the block bounds.
From the code in glimmer i think parent element is checked to verify if a node was removed or moved to another block.
In the inspector it was only used to check if a node is still in the dom.  It uses node.isConnected now.

Reply:
@patricklx I don't really understand this, but no, something sounds off. The bounds represents a range in the DOM, denoted by `(element, firstChild, lastChild)` where first/last child belongs to the element. You should expect to be able to find a sub-slice in `element.childNodes` that matches `(firstChild, lastChild)`. The inspector should be using this to highlight (bounding rect overlay). So something seems fishy about what you wanted, can you link to the code?

Reply:
yes, first and last child is the same .
And inspector only uses those:
https://github.com/emberjs/ember-inspector/blob/56575c08bcd3228d830ace596ebb778af2dd4824/ember_debug/libs/render-tree.js#L326

For another use case i would like to have parentElement. Or another way to get it.
Basically with modifiers i am putting them under an html element in inspector. Also all components that are under that element are also put under it. But i cannot know if the in-element was also under that element because it's disconnected